python/lsst/daf/butler/script/queryDatasets.py

```diff
@@ -225,30 +239,59 @@ def getDatasets(self) -> Iterator[DatasetRef]:
-            )
+        # Expand the collections query and include summary information.
+        query_collections_info = self.butler.collections.query_info(query_collections, include_summary=True)
+        query_collections = [c.name for c in query_collections_info]
+
+        # Only iterate over dataset types that are relevant for the query.
+        dataset_types = set(
+            self.butler.collections._filter_dataset_types(dataset_types, query_collections_info)
+        )
 
-            if (n_filtered := len(dataset_types)) != n_dataset_types:
-                _LOG.info("Filtered %d dataset types down to %d", n_dataset_types, n_filtered)
-            elif n_dataset_types == 0:
-                _LOG.info("The given dataset type, %s, is not known to this butler.", datasetTypes)
-            else:
-                _LOG.info(
-                    "Processing %d dataset type%s", n_dataset_types, "" if n_dataset_types == 1 else "s"
-                )
-
-            # Accumulate over dataset types.
-            for dt in sorted(dataset_types):
-                results = query.datasets(dt, collections=query_collections, find_first=self._find_first)
-                if self._where:
-                    results = results.where(self._where)
-                yield from results.with_dimension_records()
+        if (n_filtered := len(dataset_types)) != n_dataset_types:
+            _LOG.info("Filtered %d dataset types down to %d", n_dataset_types, n_filtered)
+        elif n_dataset_types == 0:
+            _LOG.info("The given dataset type, %s, is not known to this butler.", datasetTypes)
+        else:
+            _LOG.info("Processing %d dataset type%s", n_dataset_types, "" if n_dataset_types == 1 else "s")
+
+        # Accumulate over dataset types.
+        limit = self._limit
+        warn_limit = False
+        unlimited = True if limit == 0 else False
+        if limit < 0:
+            warn_limit = True
+            limit = abs(limit) + 1  # +1 to tell us we hit the limit.
+        for dt in sorted(dataset_types):
+            kwargs: dict[str, Any] = {}
+            if self._where:
+                kwargs["where"] = self._where
+            if not unlimited:
+                kwargs["limit"] = limit
+            _LOG.debug("Querying dataset type %s with %s", dt, kwargs)
+            results = self.butler.query_datasets(
+                dt,
+                collections=query_collections,
+                find_first=self._find_first,
+                with_dimension_records=True,
+                order_by=self._order_by,
+                **kwargs,
+            )
+            if not unlimited:
+                limit -= len(results)
+                if warn_limit and limit == 0 and results:
+                    # We asked for one too many so must remove that from
+                    # the list.
+                    results.pop(-1)
+            _LOG.debug("Got %d results for dataset type %s", len(results), dt)
+            yield from results
```

Comment:
Are we actually getting anything out of lazy iteration here, or do all callers eventually just put the results into regular in-memory container anyway?  For at least the case where we turn it into an Astropy table for formatting, I'm sure it'd be better to pass a `list[DatasetRef]` so Astropy can know the number of rows in advance and allocate up front accordingly.

Reply:
Yielding seemed nice given the looping over dataset types meaning that we don't require we keep everything in memory. I have been wondering why the command line doesn't try to dump the tables for each dataset type as they arrive from the system rather than waiting until the queries complete. If it yielded `list[DatasetRef]` for each dataset type then that would be more flexible.

Reply:
I made this change to yield the results of each dataset type and then yield each table immediately. Much more responsive.